modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/SqlRowHandler.java

```diff
@@ -107,12 +107,16 @@ public RowWrapper concat(RowWrapper left, RowWrapper right) {
     @Override
     public RowWrapper map(RowWrapper row, int[] mapping) {
         Object[] fields = new Object[mapping.length];
+        Builder schemaBuilder = RowSchema.builder();
+        List<TypeSpec> originalFields = row.rowSchema().fields();
 
         for (int i = 0; i < mapping.length; i++) {
-            fields[i] = row.get(mapping[i]);
+            int mappedIdx = mapping[i];
+            fields[i] = row.get(mappedIdx);
+            schemaBuilder.addField(originalFields.get(mappedIdx));
         }
 
-        return new ObjectsArrayRowWrapper(row.rowSchema(), fields);
+        return new ObjectsArrayRowWrapper(schemaBuilder.build(), fields);
```

Comment:
It's better to build schema single time for all rows.
My suggestion (may be @lowka or you have better ideas)
Add SqlRowHandler.schema(row) method
Add additional parameter to map method and pass target schema.

In splitMerge, for example, get schema of the first row and build target schema.
And do the same in flushTuples for the UPDATE operation :thinking: 

p.s. may be it's also worth to mute this failure  https://ci.ignite.apache.org/viewLog.html?buildId=8657663&buildTypeId=ApacheIgnite3xGradle_Test_RunAllTests&fromSakuraUI=true
in this PR

Reply:
Update. There is no need for SqlRowHandler.schema(row) method, just pass source schema in constructor 
```
        RelDataType rowType = rel.getInput().getRowType();
        RowSchema rowSchema = rowSchemaFromRelTypes(RelOptUtil.getFieldTypeList(rowType));
```

Reply:
Why do we have this method here, but not in RowFactory, 
where we could validate given row schema and desired new row schema?

This change will also help to avoid schema creation on every map() method call.
WDYT?